crates/wasi-common/src/snapshots/wasi_snapshot_preview1.rs

```diff
@@ -0,0 +1,1448 @@
+        let file = entry
+            .as_descriptor(
+                types::Rights::FD_READ | types::Rights::FD_SEEK,
+                types::Rights::empty(),
+            )?
+            .as_file()?;
+
+        if offset > i64::max_value() as u64 {
+            return Err(Errno::Io);
+        }
+
+        let mut buf = vec![0; total_len as usize];
+        let host_nread = match file {
+            Descriptor::OsHandle(fd) => fd::pread(&fd, &mut buf, offset)?,
+            Descriptor::VirtualFile(virt) => virt.pread(&mut buf, offset)?,
+            _ => {
+                unreachable!(
+                    "implementation error: fd should have been checked to not be a stream already"
+                );
+            }
+        };
+
+        let mut buf_offset = 0;
+        let mut left = host_nread;
+        for slice in slices {
+            if left == 0 {
+                break;
+            }
+            let vec_len = std::cmp::min(slice.len(), left);
+            slice[..vec_len].copy_from_slice(&buf[buf_offset..buf_offset + vec_len]);
+            buf_offset += vec_len;
+            left -= vec_len;
+        }
+        let host_nread = host_nread.try_into()?;
+
+        trace!("     | *nread={:?}", host_nread);
+
+        Ok(host_nread)
+    }
+
+    fn fd_prestat_get(&self, fd: types::Fd) -> Result<types::Prestat> {
+        trace!("fd_prestat_get(fd={:?})", fd);
+
+        // TODO: should we validate any rights here?
+        let fe = unsafe { self.get_entry(fd)? };
+        let po_path = fe.preopen_path.as_ref().ok_or(Errno::Notsup)?;
+        if fe.file_type != types::Filetype::Directory {
+            return Err(Errno::Notdir);
+        }
+
+        let path = path::from_host(po_path.as_os_str())?;
+        let prestat = types::PrestatDir {
+            pr_name_len: path.len().try_into()?,
+        };
+        Ok(types::Prestat::Dir(prestat))
+    }
+
+    fn fd_prestat_dir_name(
+        &self,
+        fd: types::Fd,
+        path: GuestPtr<u8>,
+        path_len: types::Size,
+    ) -> Result<()> {
+        trace!(
+            "fd_prestat_dir_name(fd={:?}, path={:?}, path_len={})",
+            fd,
+            path,
+            path_len
+        );
+
+        // TODO: should we validate any rights here?
+        let fe = unsafe { self.get_entry(fd)? };
+        let po_path = fe.preopen_path.as_ref().ok_or(Errno::Notsup)?;
+        if fe.file_type != types::Filetype::Directory {
+            return Err(Errno::Notdir);
+        }
+
+        let host_path = path::from_host(po_path.as_os_str())?;
+        let host_path_len = host_path.len().try_into()?;
+
+        if host_path_len > path_len {
+            return Err(Errno::Nametoolong);
+        }
+
+        trace!("     | (path_ptr,path_len)='{}'", host_path);
+
+        let mut bc = GuestBorrows::new();
+        let path_arr = path.as_array(host_path_len);
+        let slice = path_arr.as_raw(&mut bc)?;
+        unsafe { &mut *slice }.copy_from_slice(host_path.as_bytes());
```

Comment:
This idiom is showing up in a few places ("given this slice write this slice"), so it might be good to encapsulate this in a safe function perhaps?

Reply:
Agreed, this could be a method on GuestPtr<[T]>

Reply:
When you say "encapsulate in a safe function" do you mean as a helper in `wasi-common` or a method of `GuestPtr<'_, [T]>`? I'm partial to the latter approach, and agreed that a method like this would be most useful!

Reply:
FWIW I've posted https://github.com/bytecodealliance/wasmtime/pull/1367 to cover this.